File: PythonNCAA/scripts/_analyze_totals.py
#!/usr/bin/env python3
# scripts/_analyze_totals.py — Analyze totals pick history
import json

h = json.loads(open("data/history.json").read())

ow, ol, op = 0, 0, 0
over_w, over_l, under_w, under_l = 0, 0, 0, 0
diff_buckets = {"0-2": [], "2-4": [], "4-6": [], "6-8": [], "8+": []}
p_ou_buckets = {"0.50-0.55": [], "0.55-0.60": [], "0.60-0.65": [], "0.65-0.70": [], "0.70+": []}
total_picks, total_pass = 0, 0

for r in (h.get("runs") or []):
    if r.get("burnIn"):
        continue
    for g in (r.get("games") or []):
        if g.get("oPick") in ("PASS", None, ""):
            total_pass += 1
            continue
        total_picks += 1
        if not g.get("oResult"):
            continue
        if g["oResult"] == "PUSH":
            op += 1
            continue
        v = 1 if g["oResult"] == "WIN" else 0
        if g["oResult"] == "WIN": ow += 1
        else: ol += 1
        if g["oPick"] == "OVER":
            if v: over_w += 1
            else: over_l += 1
        elif g["oPick"] == "UNDER":
            if v: under_w += 1
            else: under_l += 1
        td = abs(g.get("tDiff") or g.get("cleanTDiff") or 0)
        if td <= 2: diff_buckets["0-2"].append(v)
        elif td <= 4: diff_buckets["2-4"].append(v)
        elif td <= 6: diff_buckets["4-6"].append(v)
        elif td <= 8: diff_buckets["6-8"].append(v)
        else: diff_buckets["8+"].append(v)
        pou = g.get("pOU", 0)
        if pou < 0.55: p_ou_buckets["0.50-0.55"].append(v)
        elif pou < 0.60: p_ou_buckets["0.55-0.60"].append(v)
        elif pou < 0.65: p_ou_buckets["0.60-0.65"].append(v)
        elif pou < 0.70: p_ou_buckets["0.65-0.70"].append(v)
        else: p_ou_buckets["0.70+"].append(v)

print(f"TOTALS OVERALL: {ow}-{ol}-{op} ({ow / (ow + ol) * 100:.1f}%)")
print(f"Total picks made: {total_picks} / Passed: {total_pass}")
print(f"Units (flat -110): {(ow - ol * 1.1):.1f}u")
print()
if over_w + over_l > 0:
    print(f"OVER picks:  {over_w}-{over_l} ({over_w / (over_w + over_l) * 100:.1f}%)")
if under_w + under_l > 0:
    print(f"UNDER picks: {under_w}-{under_l} ({under_w / (under_w + under_l) * 100:.1f}%)")

print("\nBy total diff:")
for k, v in diff_buckets.items():
    wins = sum(v)
    t = len(v)
    if t > 0:
        print(f"  {k}: {wins}-{t - wins} ({wins / t * 100:.1f}%)")

print("\nBy pOU:")
for k, v in p_ou_buckets.items():
    wins = sum(v)
    t = len(v)
    if t > 0:
        print(f"  {k}: {wins}-{t - wins} ({wins / t * 100:.1f}%)")

over_count = sum(1 for r in (h.get("runs") or []) if not r.get("burnIn") for g in (r.get("games") or []) if g.get("oPick") == "OVER")
under_count = sum(1 for r in (h.get("runs") or []) if not r.get("burnIn") for g in (r.get("games") or []) if g.get("oPick") == "UNDER")
print(f"\nPick bias: {over_count} OVER / {under_count} UNDER")

File: jsNCAA/scripts/_analyze_totals.mjs
import fs from "fs";
const h = JSON.parse(fs.readFileSync("data/history.json", "utf8"));

let ow = 0, ol = 0, op = 0;
let overW = 0, overL = 0, underW = 0, underL = 0;
const diffBuckets = { "0-2": [], "2-4": [], "4-6": [], "6-8": [], "8+": [] };
const pOUBuckets = { "0.50-0.55": [], "0.55-0.60": [], "0.60-0.65": [], "0.65-0.70": [], "0.70+": [] };

let totalPicks = 0, totalPass = 0;

for (const r of h.runs || []) {
  if (r.burnIn) continue;
  for (const g of r.games || []) {
    if (g.oPick === "PASS" || !g.oPick) { totalPass++; continue; }
    totalPicks++;

    if (!g.oResult) continue;
    if (g.oResult === "PUSH") { op++; continue; }

    const v = g.oResult === "WIN" ? 1 : 0;
    if (g.oResult === "WIN") ow++;
    else ol++;

    if (g.oPick === "OVER") { if (v) overW++; else overL++; }
    else if (g.oPick === "UNDER") { if (v) underW++; else underL++; }

    const td = Math.abs(g.tDiff || g.cleanTDiff || 0);
    if (td <= 2) diffBuckets["0-2"].push(v);
    else if (td <= 4) diffBuckets["2-4"].push(v);
    else if (td <= 6) diffBuckets["4-6"].push(v);
    else if (td <= 8) diffBuckets["6-8"].push(v);
    else diffBuckets["8+"].push(v);

    const pou = g.pOU || 0;
    if (pou < 0.55) pOUBuckets["0.50-0.55"].push(v);
    else if (pou < 0.60) pOUBuckets["0.55-0.60"].push(v);
    else if (pou < 0.65) pOUBuckets["0.60-0.65"].push(v);
    else if (pou < 0.70) pOUBuckets["0.65-0.70"].push(v);
    else pOUBuckets["0.70+"].push(v);
  }
}

console.log(`TOTALS OVERALL: ${ow}-${ol}-${op} (${(ow / (ow + ol) * 100).toFixed(1)}%)`);
console.log(`Total picks made: ${totalPicks} / Passed: ${totalPass}`);
console.log(`Units (flat -110): ${(ow - ol * 1.1).toFixed(1)}u`);
console.log();
if (overW + overL > 0) console.log(`OVER picks:  ${overW}-${overL} (${(overW / (overW + overL) * 100).toFixed(1)}%)`);
if (underW + underL > 0) console.log(`UNDER picks: ${underW}-${underL} (${(underW / (underW + underL) * 100).toFixed(1)}%)`);

console.log("\nBy total diff:");
for (const [k, v] of Object.entries(diffBuckets)) {
  const wins = v.filter(x => x).length;
  const t = v.length;
  if (t > 0) console.log(`  ${k}: ${wins}-${t - wins} (${(wins / t * 100).toFixed(1)}%)`);
}

console.log("\nBy pOU:");
for (const [k, v] of Object.entries(pOUBuckets)) {
  const wins = v.filter(x => x).length;
  const t = v.length;
  if (t > 0) console.log(`  ${k}: ${wins}-${t - wins} (${(wins / t * 100).toFixed(1)}%)`);
}

// Over vs Under bias check
let overCount = 0, underCount = 0;
for (const r of h.runs || []) {
  if (r.burnIn) continue;
  for (const g of r.games || []) {
    if (g.oPick === "OVER") overCount++;
    else if (g.oPick === "UNDER") underCount++;
  }
}
console.log(`\nPick bias: ${overCount} OVER / ${underCount} UNDER`);
